### DISCUSSION TITLE
Future direction Jakarta Data and Panache Repositories

### DISCUSSION THREAD

**Discussion by wernerjacobs** [quarkusio/quarkus#44303] [Community]
Which direction are you going with Panache? Is it the idea that it will align with the Jakarta Data or will it stay a separate option for persistence? Jakarta Data takes the route of allowing one repository for a group of related entities, which I  prefer.  Or are both (Jakarta Data and Panache Repository) supported at the same time as long as they do not support the same set of features? It would be nice to hear what are your thoughts on the matter. If we would start a new project, which persistence abstract layer would you recommend  or would you suggest using both at the same time? As I understand projection queries are not a part of Jakarta Data yet.

**Comment by cescoffier** [comment]
\CC @FroMage @yrodiere @maxandersen

**Reply by yrodiere** [reply]
cc @gavinking as well -- see [my answer below](https://github.com/quarkusio/quarkus/discussions/44303#discussioncomment-11223197), maybe you'll want to add something.

**Comment by yrodiere** [comment]
To clarify for others, you can already use Jakarta Data in Quarkus -- though documentation [needs work](https://github.com/quarkusio/quarkus/issues/44278).

@FroMage is currently working on "Panache 2" which should unify Jakarta Data, Hibernate ORM's native `@HQL` annotations and similar, and provide clearer separation between `Session` and `StatelessSession`. Last time we talked, we were supposed to have a look at his work when it's ready, and iterate from there. I'll let him elaborate on the approach he ended up taking and the current status.

One thing I'd like to stress is that Jakarta Data in JPA is fundamentally about [`StatelessSession`](https://docs.jboss.org/hibernate/orm/6.6/userguide/html_single/Hibernate_User_Guide.html#_statelesssession) -- meaning no dirty checking and no implicit lazy-loading in particular. It's a fine approach for many applications, but is also quite different from what you'd do in "traditional" JPA applications. Clarifying the differences between `EntityManager`/`Session` and `EntityAgent` (coming in JPA 4)/`StatelessSession`, and explaining when and how to use either solution, will be one of the challenges we'll have to tackle with Panache 2.

**Reply by FroMage** [reply]
Not much to add to that, that's all correct :)

**Reply by gavinking** [reply]
> Jakarta Data in JPA is fundamentally about [StatelessSession](https://docs.jboss.org/hibernate/orm/6.6/userguide/html_single/Hibernate_User_Guide.html#_statelesssession)

This is true for Data 1.0, but likely won't be true in future versions. See https://github.com/jakartaee/data/issues/470

**Comment by gavinking** [comment]
> Which direction are you going with Panache? Is it the idea that it will align with the Jakarta Data or will it stay a separate option for persistence? Jakarta Data takes the route of allowing one repository for a group of related entities, which I prefer. Or are both (Jakarta Data and Panache Repository) supported at the same time as long as they do not support the same set of features?

We've had quite a few discussion about this, and the plan is to achieve a good degree of alignment, without losing the distinct programming model that's in Panache ("active record" rather than "repository", essentially).

@FroMage is currently working on untying this knot.

> As I understand projection queries are not a part of Jakarta Data yet.

They're not _yet_ in the spec (See https://github.com/jakartaee/data/issues/539) but they're already supported in Hibernate Data Repositories https://docs.jboss.org/hibernate/orm/6.6/repositories/html_single/Hibernate_Data_Repositories.html#query-method.
